data_model/src/events/mod.rs

```diff
@@ -209,6 +209,57 @@ impl EventFilter for TriggeringEventFilterBox {
     }
 }
 
+mod conversions {
+    use super::prelude::*;
+
+    macro_rules! last_tt {
+        ($last:tt) => {
+            $last
+        };
+        ($head:tt $($tail:tt)+) => {
+            last_tt!($($tail)*)
+        };
+    }
+
+    // chain multiple conversions into one
+    macro_rules! impl_from_via_path {
+        ($($initial:ty $(=> $intermediate:ty)*),+ $(,)?) => {
+            $(
+                impl From<$initial> for last_tt!($($intermediate)*) {
+                    fn from(filter: $initial) -> Self {
+                        $(
+                            let filter: $intermediate = filter.into();
+                        )*
+                        filter
+                    }
+                }
+            )+
+        };
+    }
+
+    impl_from_via_path! {
+        PeerEventFilter             => DataEventFilter => EventFilterBox,
+        DomainEventFilter           => DataEventFilter => EventFilterBox,
+        AccountEventFilter          => DataEventFilter => EventFilterBox,
+        AssetEventFilter            => DataEventFilter => EventFilterBox,
+        AssetDefinitionEventFilter  => DataEventFilter => EventFilterBox,
+        TriggerEventFilter          => DataEventFilter => EventFilterBox,
+        RoleEventFilter             => DataEventFilter => EventFilterBox,
+        ConfigurationEventFilter    => DataEventFilter => EventFilterBox,
+        ExecutorEventFilter         => DataEventFilter => EventFilterBox,
+
+        PeerEventFilter             => DataEventFilter => TriggeringEventFilterBox,
```

Comment:
IMO we should remove `TriggeringEventFilterBox` and have only `EventFilterBox`. Is there a good reason why we have it?

Reply:
We have it to enforce on type level that `TriggerCompletedEvent` can't trigger a trigger to run to prevent infinite loops (see #2885)

Reply:
ok, got it, but we don't have to prevent it on type level

Reply:
Yeah, that would work. Can also reject transactions registering such a trigger for a bit better UX